File: src/rohrpost/merge.py
"""Three-way field merge for sync (spec §8.2, §8.3).

Sync is a three-way merge and a three-way merge needs a base. Given the shadow
snapshot (``base``, the remote's field values as of the last successful sync),
the folded ticket (``local``), and the live remote (``remote``), resolve each
mapped field:

| Condition                | Action                                  |
|--------------------------|-----------------------------------------|
| ``local == remote``      | nothing                                 |
| ``local == base``        | remote changed → take remote            |
| ``remote == base``       | local changed → push local              |
| all three differ         | conflict → apply policy                 |

Set fields compose additions and removals instead of falling through scalar
conflict handling. Prose bodies are merged with a genuine three-way text merge
via ``git merge-file`` (§8.3), keeping conflict markers on collision.
"""

from __future__ import annotations

import subprocess
import tempfile
from collections.abc import Mapping
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

#: Conflict resolution policies (spec §8.2). ``flag`` is the default.
Policy = Literal["flag", "local", "remote"]

#: The field name that holds free-form prose and therefore gets a real text merge.
BODY_FIELD: str = "body"

#: Set-valued fields use three-way set algebra, not scalar conflict handling.
SET_FIELDS: frozenset[str] = frozenset({"labels"})


@dataclass(frozen=True, slots=True)
class FieldConflict:
    """One field where local, remote and base all differ."""

    field: str
    local: object
    remote: object
    merged: object | None = None


@dataclass(frozen=True, slots=True)
class MergeResult:
    """The outcome of a three-way merge over all mapped fields.

    ``remote_won`` are field values to apply locally (append ``set`` events with
    actor ``remote/<name>``); ``local_won`` are field values to push to the
    remote; ``conflicts`` are fields that need human resolution under ``flag``.
    The merged body (if any) lands in ``remote_won`` or ``local_won`` exactly
    like any other field.
    """

    remote_won: dict[str, object] = field(default_factory=dict)
    local_won: dict[str, object] = field(default_factory=dict)
    conflicts: list[FieldConflict] = field(default_factory=list)
    resolved: list[FieldConflict] = field(default_factory=list)

    @property
    def clean(self) -> bool:
        return not self.conflicts


def merge_text(base: str, local: str, remote: str) -> tuple[str, bool]:
    """Three-way text merge via ``git merge-file``.

    Returns ``(merged_text, had_conflict)``. On conflict the markers (``<<<<<<<``
    … ``>>>>>>>``) are left in the text for the caller to flag. Falls back to the
    local text if git is unavailable, so sync degrades rather than crashes.
    """
    try:
        with tempfile.TemporaryDirectory() as tmp:
            base_p, local_p, remote_p = (Path(tmp) / n for n in ("base", "local", "remote"))
            # UTF-8 with no newline translation, never the platform defaults: on
            # Windows those are cp1252 and \n→\r\n, which would crash on
            # emoji/CJK bodies and leak CRLF into the merged text.
            base_p.write_text(base, encoding="utf-8", newline="")
            local_p.write_text(local, encoding="utf-8", newline="")
            remote_p.write_text(remote, encoding="utf-8", newline="")
            proc = subprocess.run(
                ["git", "merge-file", "-p", str(local_p), str(base_p), str(remote_p)],
                capture_output=True,
                check=False,
            )
    except FileNotFoundError:
        return local, local != remote
    # git merge-file: 0 = clean, 1 = conflicts, >1 = error/usage.
    if proc.returncode > 1:
        return local, local != remote
    merged = proc.stdout.decode("utf-8", errors="replace")
    return merged, proc.returncode == 1


def three_way(
    base: Mapping[str, object],
    local: Mapping[str, object],
    remote: Mapping[str, object],
    *,
    policy: Policy = "flag",
    body_field: str = BODY_FIELD,
    set_fields: frozenset[str] = SET_FIELDS,
) -> MergeResult:
    """Resolve every mapped field three-way.

    ``base``/``local``/``remote`` are dicts of the mapped field values (already
    translated to the local vocabulary). Fields present on only one side are
    treated as ``base``-absent changes.
    """
    result = MergeResult()
    for name in sorted(base.keys() | local.keys() | remote.keys()):
        b = base.get(name)
        lv = local.get(name)
        rv = remote.get(name)
        if name in set_fields and _all_sets(b, lv, rv):
            _merge_set(result, name, b, lv, rv)
            continue
        if lv == rv:
            continue
        if name == body_field and _all_str(b, lv, rv):
            _merge_body(result, name, b, lv, rv, policy)
        else:
            _merge_scalar(result, name, b, lv, rv, policy)
    return result


def _merge_body(
    result: MergeResult, name: str, b: object, lv: object, rv: object, policy: Policy
) -> None:
    """Three-way text-merge a prose body field; adopt or flag per policy."""
    merged, conflict = merge_text(str(b or ""), str(lv or ""), str(rv or ""))
    if conflict:
        if policy == "flag":
            result.remote_won[name] = merged
            result.conflicts.append(FieldConflict(name, lv, rv, merged))
            return
        _apply_conflict_policy(result, name, lv, rv, policy)
        return
    if merged != lv:  # the combined text differs from local -> adopt locally
        result.remote_won[name] = merged
    if merged != rv:  # ...and from remote -> push
        result.local_won[name] = merged


def _merge_scalar(
    result: MergeResult, name: str, b: object, lv: object, rv: object, policy: Policy
) -> None:
    """Resolve a non-prose field by per-field LWW against the base."""
    if lv == b:  # only remote changed
        result.remote_won[name] = rv
    elif rv == b:  # only local changed
        result.local_won[name] = lv
    else:  # all three differ
        _apply_conflict_policy(result, name, lv, rv, policy)


def _merge_set(result: MergeResult, name: str, base: object, local: object, remote: object) -> None:
    """Compose independent additions/removals using three-way set semantics."""
    base_set = _as_set(base)
    local_set = _as_set(local)
    remote_set = _as_set(remote)
    merged = (
        (base_set - (base_set - local_set) - (base_set - remote_set))
        | (local_set - base_set)
        | (remote_set - base_set)
    )
    value = sorted(merged)
    if merged != local_set:
        result.remote_won[name] = value
    if merged != remote_set:
        result.local_won[name] = value


def _as_set(value: object) -> set[str]:
    if value is None:
        return set()
    if isinstance(value, (list, tuple, set, frozenset)):
        return {str(item) for item in value}
    return set()


def _all_sets(*values: object) -> bool:
    return all(
        value is None or isinstance(value, (list, tuple, set, frozenset)) for value in values
    )


def _all_str(*values: object) -> bool:
    return all(v is None or isinstance(v, str) for v in values)


def _apply_conflict_policy(
    result: MergeResult, name: str, local: object, remote: object, policy: Policy
) -> None:
    """Resolve a conflicting scalar field per the configured policy (§8.2)."""
    if policy == "local":
        result.resolved.append(FieldConflict(name, local, remote))
        result.local_won[name] = local
    elif policy == "remote":
        result.resolved.append(FieldConflict(name, local, remote))
        result.remote_won[name] = remote
    else:  # flag — leave for a human; `rp resolve` clears it
        result.conflicts.append(FieldConflict(name, local, remote))


__all__ = [
    "BODY_FIELD",
    "SET_FIELDS",
    "FieldConflict",
    "MergeResult",
    "Policy",
    "merge_text",
    "three_way",
]

File: cpp/src/merge.cpp
#include "rohrpost/merge.hpp"

#include "rohrpost/entropy.hpp"
#include "rohrpost/io.hpp"
#include "rohrpost/subprocess.hpp"

#include <algorithm>
#include <chrono>
#include <exception>
#include <filesystem>
#include <format>
#include <optional>
#include <set>
#include <string>
#include <system_error>
#include <utility>

namespace rp::merge {
namespace {

namespace fs = std::filesystem;

Json get_or_null(const Json& obj, const std::string& name) {
    const auto it = obj.find(name);
    return it == obj.end() ? Json() : *it;
}

bool is_setlike(const Json& v) { return v.is_null() || v.is_array(); }
bool is_strlike(const Json& v) { return v.is_null() || v.is_string(); }

std::set<std::string> as_set(const Json& v) {
    std::set<std::string> out;
    if (v.is_array()) {
        for (const auto& item : v) out.insert(json::py_str(item));
    }
    return out;
}

Json sorted_json(const std::set<std::string>& s) {
    Json arr = Json::array();
    for (const auto& item : s) arr.push_back(item);
    return arr;
}

void apply_conflict_policy(MergeResult& r, const std::string& name, const Json& local, const Json& remote, Policy policy) {
    if (policy == Policy::Local) {
        r.resolved.push_back(FieldConflict{name, local, remote, std::nullopt});
        r.local_won[name] = local;
    } else if (policy == Policy::Remote) {
        r.resolved.push_back(FieldConflict{name, local, remote, std::nullopt});
        r.remote_won[name] = remote;
    } else {
        r.conflicts.push_back(FieldConflict{name, local, remote, std::nullopt});
    }
}

std::string str_or_empty(const Json& v) {
    // `str(b or "")`: None/empty -> "", else the string itself.
    if (v.is_null()) return "";
    return json::py_str(v);
}

void merge_body(MergeResult& r, const std::string& name, const Json& b, const Json& lv, const Json& rv, Policy policy) {
    auto [merged, conflict] = merge_text(str_or_empty(b), str_or_empty(lv), str_or_empty(rv));
    if (conflict) {
        if (policy == Policy::Flag) {
            r.remote_won[name] = merged;
            r.conflicts.push_back(FieldConflict{name, lv, rv, Json(merged)});
            return;
        }
        apply_conflict_policy(r, name, lv, rv, policy);
        return;
    }
    if (!json::py_equal(Json(merged), lv)) r.remote_won[name] = merged;
    if (!json::py_equal(Json(merged), rv)) r.local_won[name] = merged;
}

void merge_scalar(MergeResult& r, const std::string& name, const Json& b, const Json& lv, const Json& rv, Policy policy) {
    if (json::py_equal(lv, b)) r.remote_won[name] = rv;
    else if (json::py_equal(rv, b)) r.local_won[name] = lv;
    else apply_conflict_policy(r, name, lv, rv, policy);
}

void merge_set(MergeResult& r, const std::string& name, const Json& base, const Json& local, const Json& remote) {
    const auto base_set = as_set(base);
    const auto local_set = as_set(local);
    const auto remote_set = as_set(remote);
    std::set<std::string> merged;
    // (base - (base - local) - (base - remote)) | (local - base) | (remote - base)
    for (const auto& item : base_set) {
        if (local_set.contains(item) && remote_set.contains(item)) merged.insert(item);
    }
    for (const auto& item : local_set) {
        if (!base_set.contains(item)) merged.insert(item);
    }
    for (const auto& item : remote_set) {
        if (!base_set.contains(item)) merged.insert(item);
    }
    const Json value = sorted_json(merged);
    if (merged != local_set) r.remote_won[name] = value;
    if (merged != remote_set) r.local_won[name] = value;
}

}  // namespace

std::pair<std::string, bool> merge_text(const std::string& base, const std::string& local, const std::string& remote) {
    std::error_code ec;
    const fs::path tmp = fs::temp_directory_path(ec) / io::path_from_utf8(std::format("rohrpost-merge-{:016x}", entropy::randbits(64)));
    if (ec || !fs::create_directories(tmp, ec)) return {local, local != remote};
    const fs::path base_p = tmp / "base";
    const fs::path local_p = tmp / "local";
    const fs::path remote_p = tmp / "remote";
    try {
        io::write_file(base_p, base);
        io::write_file(local_p, local);
        io::write_file(remote_p, remote);
    } catch (const std::exception&) {
        fs::remove_all(tmp, ec);
        return {local, local != remote};
    }
    auto proc = subprocess::run({"git", "merge-file", "-p", io::path_str(local_p), io::path_str(base_p), io::path_str(remote_p)},
                                std::chrono::minutes(5));
    fs::remove_all(tmp, ec);
    // git merge-file: 0 = clean, 1 = conflicts, >1 = error/usage.
    if (!proc || proc->returncode > 1 || proc->returncode < 0) return {local, local != remote};
    return {proc->stdout_bytes, proc->returncode == 1};
}

MergeResult three_way(const Json& base, const Json& local, const Json& remote, Policy policy) {
    MergeResult result;
    std::set<std::string> names;
    for (const auto* src : {&base, &local, &remote}) {
        if (src->is_object()) {
            for (const auto& [k, v] : src->items()) names.insert(k);
        }
    }
    for (const auto& name : names) {
        const Json b = get_or_null(base, name);
        const Json lv = get_or_null(local, name);
        const Json rv = get_or_null(remote, name);
        if (name == "labels" && is_setlike(b) && is_setlike(lv) && is_setlike(rv)) {
            merge_set(result, name, b, lv, rv);
            continue;
        }
        if (json::py_equal(lv, rv)) continue;
        if (name == kBodyField && is_strlike(b) && is_strlike(lv) && is_strlike(rv)) merge_body(result, name, b, lv, rv, policy);
        else merge_scalar(result, name, b, lv, rv, policy);
    }
    return result;
}

}  // namespace rp::merge
